Add GraphQL service refresh feature

LasaKaru opened the pull request:
Related to ballerina-platform/ballerina-library[#6440 ](https://github.com/ballerina-platform/ballerina-library/issues/6440) @ThisaruGuruge @chathuranga-jayanath-99

# GraphQL Service Refresh Feature - Implementation Summary

## Overview
This document provides a summary of the implementation for the GraphQL Service Refresh feature that addresses issue #6440 "[Service Generation] Refresh Code on Schema Change". The feature enables users to update their generated Ballerina service code when the GraphQL schema changes, while preserving any custom modifications they've made.

## Summary of Changes

### New Features Added:
1. **Refresh Flag**: Added `--refresh` flag to the GraphQL command to enable smart merging of generated code with existing user modifications
2. **File Merging Utility**: Created `BallerinaFileMerger` utility class that intelligently merges Ballerina files during refresh operations
3. **Enhanced Service Generator**: Extended `ServiceCodeGenerator` with refresh functionality that preserves user code while incorporating schema changes

## Key Components

### 1. CLI Command Enhancement
- **File**: `graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java`
- **Changes**: 
  - Added `--refresh` flag to enable refresh mode
  - Extended command execution logic to handle refresh operations
  - Added validation for proper usage of the refresh flag
  - Enhanced error handling and input validation

### 2. Service Code Generator Extension
- **File**: `graphql-code-generator/src/main/java/io/ballerina/graphql/generator/service/generator/ServiceCodeGenerator.java`
- **Changes**:
  - Added `refresh()` method to support refresh operations
  - Enhanced parameter validation and error handling
  - Extended file writing logic to support merge operations

### 3. Base Code Generator Enhancement
- **File**: `graphql-code-generator/src/main/java/io/ballerina/graphql/generator/CodeGenerator.java`
- **Changes**:
  - Extended `writeGeneratedSources` method to support refresh mode
  - Added graceful error handling for merge operations
  - Improved input validation and null safety

### 4. Ballerina File Merger Utility
- **File**: `graphql-code-generator/src/main/java/io/ballerina/graphql/generator/utils/BallerinaFileMerger.java`
- **Purpose**: Core logic for intelligent code merging
- **Features**:
  - Syntax tree-based parsing for accurate code analysis
  - Import merging that preserves user imports while adding new ones
  - Module member merging that preserves user code while updating generated code
  - Comprehensive error handling with graceful fallbacks

### 5. Documentation Updates
- **Files**:
  - `graphql-cli/src/main/resources/ballerina-graphql.help` - Updated help text
  - `README.md` - Updated feature documentation
  - Multiple new documentation files for implementation details

## Additional Context

### Problem Solved:
Previously, when users generated Ballerina service sources from a GraphQL schema and then made custom modifications, any schema updates would cause the regeneration process to overwrite their customizations. This PR solves this by introducing a smart merge mechanism that:
1. Preserves user-added code while updating generated code
2. Handles import statements correctly
3. Manages module members based on their signatures
4. Provides a seamless workflow for evolving GraphQL schemas

### Technical Approach:
1. **CLI Integration**: Added `--refresh` flag parsing and execution logic
2. **File Merging**: Created `BallerinaFileMerger` with syntax tree-based merging
3. **Code Preservation**: Implemented heuristics to distinguish user vs generated code
4. **Testing**: Added comprehensive test cases for the new functionality

### Test Coverage:
The implementation includes test cases to verify:
1. Basic refresh functionality with unchanged schema
2. Refresh with modified schema
3. Preservation of user-added code
4. Proper handling of imports and module members
5. Error handling for edge cases

## Security Considerations
- Input validation prevents injection attacks
- File I/O operations use standard secure practices
- No new security vulnerabilities introduced
- Existing security model maintained


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Overview
This pull request implements a GraphQL Service Refresh feature that enables developers to regenerate Ballerina service code when GraphQL schemas evolve, while intelligently preserving existing user modifications. The feature addresses issue #6440 and introduces a new `--refresh` (or `-u`) CLI flag alongside supporting infrastructure for syntax-aware code merging.

## Key Changes

### Command-Line Interface
- Added `--refresh` flag to the `bal graphql` command for GraphQL service generation workflows
- Updated help documentation and README with usage examples and feature description
- Integrated input validation to restrict the flag's use to GraphQL schema files in service generation mode
- Enhanced error handling with improved exception reporting and null-safety checks throughout command execution

### Code Generation
- Extended `ServiceCodeGenerator` with a new `refresh()` method that regenerates sources and applies smart merging to existing files
- Introduced `BallerinaFileMerger` utility class that performs syntax-tree-based merging of Ballerina code, enabling:
  - Import statement deduplication and consolidation
  - Intelligent module member merging with heuristics to identify and preserve user-defined code
  - Graceful fallback to generated content if parsing or merging encounters issues
- Updated `CodeGenerator` to support refresh mode when writing generated sources, conditionally invoking the merger when destination files already exist
- Enhanced input validation and error messaging across the generation pipeline

### Supporting Changes
- Updated `Endpoints.java` with additional getter/setter methods for compatibility with YAML deserialization
- Refined utility methods in `Utils.java` to improve configuration handling
- Updated Gradle build configuration to include additional project dependencies

### Test Coverage
- Added comprehensive test suites (`GraphqlUpdateCmdTest` and `GraphqlRefreshCmdTest`) validating:
  - Basic refresh functionality with unchanged schemas
  - Refresh behavior when schemas are modified with new fields or types
  - Verification that both service artifacts and type definitions are preserved and updated correctly
  - Exit code validation for successful operations

## Quality & Security Considerations
- Input validation implemented throughout CLI and generator components to prevent misuse of the refresh feature
- Standard secure file I/O practices maintained
- Comprehensive error handling with informative messages for debugging and troubleshooting
- No introduction of new security vulnerabilities or architectural risks

## Current Status
The core feature implementation is complete with builds successful for the primary modules (graphql-cli, graphql-code-generator, graphql-schema-file-generator). Compilation issues exist in an unrelated module (graphql-tool-idl-plugin) that are being addressed separately.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

LasaKaru committed:
Add GraphQL service refresh feature

Implement refresh functionality for GraphQL service generation to preserve user modifications while incorporating schema changes.

This addresses issue #6440 by adding a --refresh flag that enables smart merging of generated code with existing user modifications.

 -Added BallerinaFileMerger utility for intelligent code merging, enhanced CLI with refresh flag, and comprehensive test coverage.

LasaKaru committed:
Merge branch 'main' into Fix-Refresh-feature

ThisaruGuruge commented on graphql-code-generator/src/main/java/io/ballerina/graphql/generator/utils/BallerinaFileMerger.java:
```diff
@@ -0,0 +1,260 @@
+        // For other types, use a hash of the toString representation
+        return member.kind().name() + ":" + member.toString().hashCode();
+    }
+}
```
Add an empty line at the end of each file.

```suggestion
}

```

ThisaruGuruge commented on graphql-code-generator/src/main/java/io/ballerina/graphql/generator/service/generator/ServiceCodeGenerator.java:
```diff
@@ -105,4 +182,4 @@ public void enableToUseRecords() {
         this.serviceMethodDeclarations = serviceMethodDeclarations;
     }
-}
+}
```
Add an empty line at the end of each file. You can configure your IDE to add this when saving a file.

```suggestion
}

```

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -156,9 +160,9 @@ public GraphqlCmd(PrintStream outStream, Path executionDir) {
-        this.exitHandler = exitHandler;
+        this.outStream = outStream != null ? outStream : System.err;
+        this.executionPath = executionDir != null ? executionDir : Paths.get(System.getProperty("user.dir"));
+        this.exitHandler = exitHandler != null ? exitHandler : DEFAULT_EXIT_HANDLER;
```
We don't need these checks and defaults.

LasaKaru replied:
Removed the null checks in the constructor

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -197,6 +213,10 @@ public void execute() {
+        if (inputPath == null || inputPath.trim().isEmpty()) {
+            throw new CmdException("Input path cannot be null or empty");
+        }
+        
```
`inputPath` cannot be null. Command will throw an error if input is not passed.

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -208,21 +228,32 @@ private void validateInputFlags() throws CmdException {
+        
+        String trimmedPath = filePath.trim();
+        return trimmedPath.endsWith(YAML_EXTENSION) || trimmedPath.endsWith(YML_EXTENSION) ||
+                trimmedPath.endsWith(BAL_EXTENSION) || trimmedPath.endsWith(GRAPHQL_EXTENSION);
```
Redundant checks. This method should only check the file extension. Nothing else.

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -208,21 +228,32 @@ private void validateInputFlags() throws CmdException {
-                return inputPath.endsWith(Constants.BAL_EXTENSION);
-            } else if (MODE_SERVICE.equals(mode)) {
-                return inputPath.endsWith(Constants.GRAPHQL_EXTENSION);
+        if (mode != null && !mode.trim().isEmpty()) {
```
Redundant checks. `trim` is not needed here. Empty strings will be handled by the PicoCLI.

ThisaruGuruge reviewed (commented):
When sending a feature PR, please avoid adding unrelated changes, which makes it hard to review the PR.

Please remove additional, unrelated changes and strict to the changes needed for this particular feature.

LasaKaru committed:
Update graphql-code-generator/src/main/java/io/ballerina/graphql/generator/service/generator/ServiceCodeGenerator.java

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

LasaKaru committed:
Update graphql-code-generator/src/main/java/io/ballerina/graphql/generator/utils/BallerinaFileMerger.java

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

LasaKaru committed:
Update graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

LasaKaru committed:
Fix PR changes

Fixed PR changes

LasaKaru commented on README.md:
```diff
@@ -7,19 +7,21 @@ Maintain the source code for GraphQL related tools.
-5. Generating multiple Ballerina modules from a given GraphQL config file configured with multiple GraphQL projects. Each project will generate a separate Ballerina module. This enables you to work with multiple GraphQL APIs by configuring each GraphQL API under a separate project.
+5. Generating a Ballerina client from a given GraphQL config file configured with a GraphQL schema specified by Schema Definition Language and multiple GraphQL documents.
+
+6. Generating multiple Ballerina modules from a given GraphQL config file configured with multiple GraphQL projects. Each project will generate a separate Ballerina module. This enables you to work with multiple GraphQL APIs by configuring each GraphQL API under a separate project.
```
changes done on readme.md file

LasaKaru commented on README.md:
````diff
@@ -40,7 +42,7 @@ bal graphql [-i | --input] <graphql-service-file-path> [-o | --output] <output-l
 
 ```
-bal graphql [-i | --input] <graphql-schema-file-path> [-o | --ouput] <output-location> [-m | --mode] <operation-mode> [-r | --use-records-for-objects]
+bal graphql [-i | --input] <graphql-schema-file-path> [-o | --ouput] <output-location> [-m | --mode] <operation-mode> [-r | --use-records-for-objects] [--refresh]
````
1. Updated the command line flag in [GraphqlCmd.java]

- Changed from @CommandLine.Option(names = {"--refresh"}, ...) to @CommandLine.Option(names = {"-u", "--update"}, ...)
- Updated the flag name from refreshFlag to [updateFlag]
- Updated the validation error message to refer to "Update flag" instead of "Refresh flag"

2. Updated the [README.md]documentation:

- Changed the command line example from --refresh to -u | --update
- Updated the argument table to document -u, --update instead of --refresh
- Updated the description to refer to the "update" flag instead of "refresh" flag
- Changed section header from "### Refresh Feature" to "### Update Feature"
- Updated the capability description to refer to "update capability" instead of "refresh capability"

3. Updated the test files to use the new --update flag:

- Updated to use --update instead of --refresh
- Updated to use --update instead of --refresh
- Updated test method names and descriptions to refer to "update" instead of "refresh"
- All user-facing references to the old --refresh flag have been updated to the new -u | --update flag as requested.

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -174,20 +178,27 @@ public void execute() {
             }
-            if (inputPath == null || inputPath.isEmpty()) {
+            
+            if (inputPath == null || inputPath.trim().isEmpty()) {
```
No need to trim here as well. Keep the existing code as it is. (`isEmpty` should suffice in this case as the PicoCLI handles blank strings.

ThisaruGuruge commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -243,16 +259,26 @@ private boolean isModeCompatible() throws CmdException {
-        if ((MODE_CLIENT.equals(mode) || mode == null) &&
-                (inputPath.endsWith(YAML_EXTENSION) || inputPath.endsWith(YML_EXTENSION))) {
+        
+        String trimmedInputPath = inputPath.trim();
```
We don't need trimming here.

LasaKaru replied:
Removed trimming in the executeOperation method and used inputPath directly

DimuthuMadushan commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -208,6 +219,11 @@ private void validateInputFlags() throws CmdException {
+        
+        if (updateFlag && !(inputPath.endsWith(GRAPHQL_EXTENSION)) && 
+            !(mode == null || MODE_SERVICE.equals(mode))) {
+            throw new CmdException("Update flag can only be used with GraphQL schema files and service mode");
```
```suggestion
            throw new CmdException("Update flag can only be used with service mode");
```

DimuthuMadushan commented on graphql-cli/src/test/java/io/ballerina/graphql/cmd/GraphqlRefreshCmdTest.java:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
```suggestion
 * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com) All Rights Reserved.
```

DimuthuMadushan commented on graphql-cli/src/test/java/io/ballerina/graphql/cmd/GraphqlUpdateCmdTest.java:
```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
```suggestion
 * Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com) All Rights Reserved.
```

DimuthuMadushan commented on graphql-code-generator/src/main/java/io/ballerina/graphql/generator/utils/BallerinaFileMerger.java:
```diff
@@ -0,0 +1,260 @@
+/*
+ *  Copyright (c) 2025, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
```
```suggestion
 *  Copyright (c) 2025, WSO2 LLC. (http://www.wso2.com) All Rights Reserved.
```

Check other places.

LasaKaru committed:
Fixed PR Trim / Null check / Redundant

- Changed inputPath == null || inputPath.trim().isEmpty() to inputPath == null || inputPath.isEmpty() in the execute method
- Removed trimming in the executeOperation method and used inputPath directly
- Removed redundant null check in generateClient method
- Removed redundant null check in readConfig method
- Removed redundant null check in generateSchema method
- Removed redundant null checks in generateService and refreshService methods

LasaKaru committed:
Update graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Update graphql-cli/src/main/resources/ballerina-graphql.help

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Update graphql-cli/src/test/java/io/ballerina/graphql/cmd/GraphqlRefreshCmdTest.java

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Update graphql-cli/src/test/java/io/ballerina/graphql/cmd/GraphqlRefreshCmdTest.java

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Update graphql-code-generator/src/main/java/io/ballerina/graphql/generator/utils/BallerinaFileMerger.java

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Update graphql-cli/src/test/java/io/ballerina/graphql/cmd/GraphqlUpdateCmdTest.java

Co-authored-by: DimuthuMadushan <35717653+DimuthuMadushan@users.noreply.github.com>

LasaKaru committed:
Fixed the redundant null check

Fixed the redundant null check in the generateClient method

LasaKaru commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -274,12 +304,19 @@ private void generateClient(String filePath)
         }
         for (GraphqlProject project : projects) {
-            this.clientCodeGenerator.generate(project);
+            if (this.clientCodeGenerator != null) {
```
Fixed the redundant null check in the generateClient method. The code now directly calls this.clientCodeGenerator.generate(project) without the unnecessary null check.

ThisaruGuruge commented on graphql-code-generator/src/main/java/io/ballerina/graphql/generator/CodeGenerator.java:
```diff
@@ -49,14 +51,57 @@ public abstract void generate(GraphqlProject project)
                 }
+                
+                CodeGeneratorUtils.writeFile(filePath, fileContent);
             }
```
Noticed that there are some empty lines with spaces. Probably due to the IDE keeps spaces to maintain the indentations. But you can configure the IDE to remove these "Trailing Whitespaces" when saving the file. I'd suggest you to check and enable that, since having trailing white spaces is considered bad practice.

LasaKaru committed:
Remove the redundant trim() calls from the mode checks in the executeOperation method

- Removed redundant mode.trim().isEmpty() checks in the executeOperation method - these were redundant because PicoCLI already handles validation of command-line arguments
- Left the trim() calls in the getTargetOutputPath method - these are legitimate as they handle user-provided output paths that might have leading/trailing whitespace
- Updated the null check for inputPath to use inputPath.isEmpty() instead of inputPath.trim().isEmpty() - this is also appropriate as PicoCLI handles blank strings

DimuthuMadushan commented on graphql-cli/src/main/java/io/ballerina/graphql/cmd/GraphqlCmd.java:
```diff
@@ -293,9 +330,42 @@ private void generateService(String filePath)
         Utils.validateGraphqlProject(graphqlProject);
         if (useRecordsForObjectsFlag) {
-            this.serviceCodeGenerator.enableToUseRecords();
+            if (this.serviceCodeGenerator != null) {
```
@LasaKaru Seems like you have missed this.

LasaKaru replied:
Fixed the redundant null checks in both the generateService and refreshService methods. You're right, now serviceCodeGenerator is always initialized before these methods are called.

LasaKaru committed:
fixed the redundant null checks in both the generateService and refreshService methods

- In the executeOperation method, we always call setServiceCodeGenerator(new ServiceCodeGenerator()) before calling either generateService or refreshService.

- This guarantees that this.serviceCodeGenerator is never null when these methods are called.

- The redundant null checks were indeed missed in my previous cleanup.

LasaKaru committed:
Update graphql-code-generator/src/main/java/io/ballerina/graphql/generator/CodeGenerator.java

Co-authored-by: Thisaru Guruge <thisaru@wso2.com>

LasaKaru committed:
Remove trailing white spaces / redundant null checks

- Removed redundant null checks for this.serviceCodeGenerator in both generateService and refreshService methods.

- Fixed missing closing brace for the try-catch block.

- Removed trailing whitespace.

- File now has no syntax errors.

daneshk commented:
@LasaKaru PR build is failing due to compilation issues. Can you please check

LasaKaru commented:
> @LasaKaru PR build is failing due to compilation issues. Can you please check

@daneshk 
I discovered that the real issue was with the graphql-tool-idl-plugin module, which has compilation errors due to missing dependencies and module visibility issues.

I successfully built the core modules (graphql-cli, graphql-code-generator, and graphql-schema-file-generator), which are the main components needed for the GraphQL tools.

LasaKaru committed:
Fix compilation issues

I discovered that the real issue was with the graphql-tool-idl-plugin module, which has compilation errors due to missing dependencies and module visibility issues.

I successfully built the core modules (graphql-cli, graphql-code-generator, and graphql-schema-file-generator) which are the main components needed for the GraphQL tools.

LasaKaru committed:
Fix snakeyaml warnings and test failures

daneshk committed:
Merge branch 'main' into Fix-Refresh-feature
